driver/src/contracts/base_contract.rs

```diff
@@ -22,26 +34,81 @@ impl BaseContract {
 
         let json: serde_json::Value = serde_json::from_str(&contents)?;
-        let abi: String = json.get("abi").ok_or("No ABI for contract")?.to_string();
+        let abi_string = json.get("abi").ok_or("No ABI for contract")?.to_string();
+        let abi = abi_string.as_bytes();
 
         let decoded_address = hex::decode(&address[2..])?;
         let contract_address: Address = Address::from(&decoded_address[..]);
-        let contract = Contract::from_json(web3.eth(), contract_address, abi.as_bytes())?;
+        let contract = Contract::from_json(web3.eth(), contract_address, abi)?;
+
+        let network_id = env::var("NETWORK_ID")?.parse()?;
+        let private_key = H256::from_str(&env::var("PRIVATE_KEY")?)?;
 
         Ok(BaseContract {
             contract,
             web3,
             event_loop,
+            abi: ethabi::Contract::load(abi)?,
+            network_id,
+            public_key: address_from_private_key(&private_key)?,
+            private_key,
         })
     }
 
-    pub fn account_with_sufficient_balance(&self) -> Option<Address> {
-        let accounts: Vec<Address> = self.web3.eth().accounts().wait().ok()?;
-        accounts
-            .into_iter()
-            .find(|&acc| match self.web3.eth().balance(acc, None).wait() {
-                Ok(balance) => !balance.is_zero(),
-                Err(_) => false,
-            })
+    pub fn send_signed_transaction<P>(
+        &self,
+        function: &str,
+        params: P,
+        options: Options,
+    ) -> Result<H256>
+    where
+        P: Tokenize,
+    {
+        let nonce = if let Some(nonce) = options.nonce {
+            nonce
+        } else {
+            self.web3
+                .eth()
+                .transaction_count(self.public_key, None)
+                .wait()?
+        };
+
+        info!("Sending tx from {} with nonce {:x}", self.public_key, nonce);
+
+        let signed_tx = self
+            .abi
+            .function(function)
+            .and_then(|function| function.encode_input(&params.into_tokens()))
+            .map(move |data| {
+                let tx = RawTransaction {
+                    nonce,
+                    to: Some(self.contract.address()),
+                    value: options.value.unwrap_or_default(),
+                    gas_price: options
+                        .gas_price
+                        .unwrap_or_else(|| U256::from(1_000_000_000)),
+                    gas: options.gas.unwrap_or_else(|| U256::from(100_000)),
+                    data,
+                };
+                tx.sign(&self.private_key, &4)
+            })?;
+
+        self.web3
+            .eth()
+            .send_raw_transaction(Bytes::from(signed_tx))
+            .wait()
+            .map_err(DriverError::from)
     }
 }
+
+fn address_from_private_key(pk: &H256) -> Result<H160> {
+    let public_key = PrivateKey::from_slice(&pk.to_vec())
+        .and_then(|pk| pk.to_public_key())
```

Comment:
I assume `to_public_key` is a function implemented for `H256`

Reply:
It's implemented on PrivateKey: https://docs.rs/clarity/0.1.23/clarity/private_key/struct.PrivateKey.html#method.to_public_key